### PR TITLE
Adds noInterrupt() and interrupt() functionality

### LINKED ISSUE #7211: noInterrupts() not implemented - can not turn off interrupts
### Board

ESP32-PICO-MINI-02

### Device Description

Adafruit Feather ESP32 V2
https://www.adafruit.com/product/5400

### Hardware Configuration

No additional hardware attached. However, to run the example sketch to demonstrate the issue, two digital pins need to be connected together somehow. This was done on a breadboard using a jumper wire.

### Version

v2.0.4

### IDE Name

Arduino IDE

### Operating System

Ubuntu

### Flash frequency

n/a

### PSRAM enabled

yes

### Upload speed

921600

### Description

The Arduino methods for disabling/enabling interrupts are no implemented:
https://www.arduino.cc/reference/en/language/functions/interrupts/nointerrupts/
https://www.arduino.cc/reference/en/language/functions/interrupts/interrupts/

Interrupts are "always on".

### Sketch

```cpp
#define OUT_PIN 14
#define IN_PIN 32

int count = 0;

void counter() {
  count++;
}

void setup() {
  Serial.begin(9600);

  pinMode(OUT_PIN, OUTPUT);
  digitalWrite(OUT_PIN, LOW);
  pinMode(IN_PIN, INPUT);

  attachInterrupt(digitalPinToInterrupt(IN_PIN), counter, RISING);

  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }

  noInterrupts();

  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }

  Serial.print("count = "); Serial.println(count);
}

void loop() {
}
```

### Debug Message

```plain
No debug messages. Sketch runs without issue. However, the expected output is 50, not 100.


count = 100
```
```


### Other Steps to Reproduce

This is the same issue reported in this older thread:
https://github.com/espressif/arduino-esp32/issues/832
that was closed claiming it's not a BSP issue.

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by SuGlider**
## Description of Change
Adds noInterrupt() and interrupt() to Arduino. 
This must be used carefully because ESP32 Arduino runs under FreeRTOS and it is easy to get Watchdog Timeout error if it is no correctly used.

## Tests scenarios
Tested with this sketch (as suggested in the issue)

``` cpp
//  To run the example sketch to demonstrate the issue, 
//  two digital pins need to be connected together somehow. 
//  This was done on a breadboard using a jumper wire.

#define OUT_PIN 14
#define IN_PIN 32

int count = 0;

void counter() {
  count++;
}

void setup() {
  Serial.begin(115200);

  pinMode(OUT_PIN, OUTPUT);
  digitalWrite(OUT_PIN, LOW);
  pinMode(IN_PIN, INPUT);

  attachInterrupt(digitalPinToInterrupt(IN_PIN), counter, RISING);

  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }

  noInterrupts();
  
  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }
  
  interrupts();

  Serial.print("count = "); Serial.println(count);
}

void loop() {
}
```

#### Expected output:
```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13864
load:0x40080400,len:3608
entry 0x400805f0
count = 50
```
wrong output (before this PR) says `count = 100`

## Related links
Fixes #7211

### CODE REVIEW

**Comment by PilnyTomas** [comment]
Tested on S3 and I'm getting 51 instead of 50. Tried various pins:
OUT  |  IN
  14   |  13
  14   |  47
  40   |  47

**Comment by caternuson** [comment]
Thanks. This seems to resolve the issue with the library here:
https://github.com/adafruit/Adafruit_VS1053_Library/issues/45

However, I'm also seeing the extra count show up in the simple test sketch. Revised test sketch:
```cpp
#define OUT_PIN 14
#define IN_PIN 32

int count = 0;

void counter() {
  count++;
}

void setup() {
  Serial.begin(9600);

  pinMode(OUT_PIN, OUTPUT);
  digitalWrite(OUT_PIN, LOW);
  pinMode(IN_PIN, INPUT);

  attachInterrupt(digitalPinToInterrupt(IN_PIN), counter, RISING);

  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }

  Serial.print("count = "); Serial.println(count);

  noInterrupts();

  for (int i=0; i<50; i++) {
    digitalWrite(OUT_PIN, HIGH);
    digitalWrite(OUT_PIN, LOW);
  }

  interrupts();

  Serial.print("count = "); Serial.println(count);
}

void loop() {
}
```

Outputs:
```
count = 50
count = 51
```

**Comment by P-R-O-C-H-Y** [comment]
I am seeing just same result as guys mentioned above.
I think it may be due to interrupt flag is still set in registers after `noInterrupts();`, but its triggered just after `interrupts();` is called.

**Review by P-R-O-C-H-Y — APPROVED** [review]
From my point of view it works as expected, there is no easy way to clear all coming interrupts before` Interrupts()` is called. As suggested you must be careful when using `noInterrupts()`, and its better to use `detachInterrupt()`.

File: cores/esp32/Arduino.h [hunk 1]
```diff
@@ -79,10 +79,9 @@
 #define degrees(rad) ((rad)*RAD_TO_DEG)
 #define sq(x) ((x)*(x))
 
-#define sei()
-#define cli()
```
Comment (me-no-dev) [inline cores/esp32/Arduino.h:83]: You could have define sei and cli instead for backwards compatibility with older code. Any reason you didn't @SuGlider?

Comment (SuGlider) [reply cores/esp32/Arduino.h:83]: As far as I know, SEI and CLI are ASM instructions used in ATMEL chips. 
Arduino official API is `interrupts()` and `noInterrupts()`
I see no reason for defining those ATMEL specific functions.

``` cpp
//interrupts is defined as:

#define interrupts() sei()
//and sei is defined as for the UNO/328p:

# define sei()  __asm__ __volatile__ ("sei" ::)
```

Comment (SuGlider) [reply cores/esp32/Arduino.h:83]: Looking ESP8266 `Arduino.h` I see that it also doesn't define `sei()` nor `cli()`.
My vote is for keeing it as defined in this PR.

**Review by PilnyTomas — APPROVED** [review]
LGTM

File: cores/esp32/Arduino.h (hunk 1 above)
Comment (PilnyTomas) [reply cores/esp32/Arduino.h:83]: `interrupts()` and `noInterrupts()` are described in [Arduino Language Reference](https://www.arduino.cc/reference/en/language/functions/interrupts/nointerrupts/) so I vote for this too.

**Comment by me-no-dev** [comment]
`sei()` and `cli()` are AVR functions present in lots of old Arduino code. I see no fault to have them there

**Comment by PilnyTomas** [comment]
> `sei()` and `cli()` are AVR functions present in lots of old Arduino code. I see no fault to have them there

Having both variants should not hurt...

**Comment by me-no-dev** [comment]
@SuGlider will merge as soon as you re-add sei and cli :)

**Pull request merged**